### ISSUE TITLE
Allow offline setup with map restore option

### ISSUE THREAD

**Issue by loopdeeloop** [organicmaps/organicmaps#4517]
**Is your feature request related to a problem? Please describe.**
Please add an option to backup and restore maps to allow offline setup.

**Describe the solution you would like**
An option to backup maps and to restore them after installing the app for the first time and in the menu later on.

**Describe alternatives you have considered**
-

**Additional context**
-

**Comment by biodranik** [comment]
What is your use case? Why do you need it? What is your platform? Why you can't download all maps again in the app? Why can't you download maps from a [mirror](https://cdn.organicmaps.app/mirrors/)?

**Comment by biodranik** [comment]
Can't you use some custom backup apps for that? I remember users mentioned it as a _bug_ in some of our issues, that maps _are backed up_ with some 3party tool.

**Comment by biodranik** [comment]
Another important point to consider: maps are constantly updated. We plan to release updates weekly. So for the offline setup, it would be better to install the actual apk and download the freshest maps data from the mirror.

**Comment by biodranik** [comment]
No, maps data won't work well or won't work at all in different app versions.

**Comment by biodranik** [comment]
The same version should work.

Can you please elaborate on that "backup option"? How other apps are implementing it?

Is your case really about backing up the old data, or is it more about "I need to install a fresh OM and fresh maps data to some device that doesn't have an internet connection"?

**Comment by biodranik** [comment]
Can you check the folder structure in this zip file, and create it manually from the mirrored OM data?

**Comment by biodranik** [comment]
That is possible in theory, there are related issues #581 #1801.

**Comment by IzzySoft** [comment]
> So for the offline setup, it would be better to install the actual apk and download the freshest maps data from the mirror.

Won't work if the device is offline – and I mean, **totally offline**. I have such a case here currently: device has no network (no SIM, no WiFi). So I've copied the data directory (with the map files) over from another device. On first start, OM still insists to download the world map (which is present in its directory) and won't let me skip that step. So I cannot use it on that device.

Yes, #581 is a good relation. It's the same app version on both devices, and no issue to update it on the offline device (using `adb install` via USB, or even copying the APK via MTP and installing/updating it via file manager) to keep the two in sync.

I still mention this here and not in 581, as I wouldn't even get to the "import" menu on first start as the "download world map" window is blocking me. Same with #1081, I won't get that far if sharing would happen inside the app.

Device in question is running Android 14, not rooted. I still can access the `Android/data` directory via ADB, so I've simply copied over the files (here: the `data/app.organicmaps/files/240326` directory which contains two files, `World.mwm` and `WorldCoasts.mwm`). No dice, OM seems to ignore them.

**Comment by biodranik** [comment]
Such a setup should work if you do it properly. Did you try to use the latest OM versions and the latest map data? Btw, installing a Google Play build won't require downloading world maps.

**Comment by IzzySoft** [comment]
I've used the data from an existing device, where OM was updated to the latest version and the map data I used did (as far as I remember) not show the need to be updated. While I rarely update *all* map data, I do update those I need. But now that I check on the "source device", indeed all 3 maps have updates. Let me try again now that the update is through…

Nope. Had to uninstall OM (as without root I no longer could access its directories in `Android/data/app.organicmaps`, which somehow meanwhile had their ownership transferred to the app and no longer belonged to `shell` – no idea when or how that happened). The recreated the directory, placed the recent files extracted from the source device after having updated the maps (`240723` directory), disconnected ADB, reinstalled and started OM. It insists on downloading the world map.

Checking the files on-device show different ownership (my copied files are still owned by `shell`, while the parent was transferred to `u0_a227`). Without root, I cannot change ownership (`Operation not permitted`). Let's see if maybe Android adjusts ownership on boot… nope, it doesn't. OM still wants me to "Download the world map" :cry:

> installing a Google Play build won't require downloading world maps.

And where would I get that? The only APK at releases is the one from `app.organicmaps.web`. I've just tried that, and it has the very same behavior. I have no Google account, and don't want one (having cut my ties to that eco-system 10 years ago for a reason) – so no way to download from there.

Wouldn't it be an option to have that screen skipped?

> Such a setup should work if you do it properly. Did you try to use the latest OM versions and the latest map data?

So that's what I just tried now, and failed. Probably due to the ownership of directories and files.

**Comment by biodranik** [comment]
Did you try to copy World*.mwm into `/sdcard/Android/data/app.organicmaps.web/files/` dir? Note that OM should be restarted to detect files.

**Comment by IzzySoft** [comment]
As outlined above: all files put to `/sdcard/Android/data/app.organicmaps.web/files/240723/`, as they were on the source device. That place contains `World.mwm`, `WorldCoasts.mwm` plus one local map of the place I live at. And not only did I restart OM, I even rebooted the device.

As it sounds that I should have placed the files directly into `files/`, I've tried that now. That seems to trigger something: on starting OM, I now get a "Disk error" stating "Unable to create folder and move files on internal device's memory or sdcard" (makes sense, as the "folder" has the wrong ownership). Backing out of the error I see the previous demand to download the worldmap.

Exiting OM, placing all files into `files/`, removing the `240723` directory, starting OM again – Yay! World map accepted! Just not the local map. Zooming in to that area, OM wants me to download it :cry:

OK, one step further. Now, how do I get the other map in without any network connection? Looking at `sdcard/Android/data/app.organicmaps.web/files` (I've used the APK from releases here this time), I see `World.mwm` and `WorldCoasts.mwm` have been moved into `240723`, but the other map remained directly in `files`. I cannot copy it into `240723` without root (because of ownership). And I cannot find any import menu, or a place to specify where files should be auto-imported from. Using a file manager to "share" it to OM has OM popping uo a message saying "Loading Bookmarks" and "The file type is not recognized by the app". That would again be the area of #581 and #1801 – but neither of the two has a solution yet.

Fun facts:

* the two files which were moved into `240723` are still owned by "shell"
* I just thought "you never know", went in via `adb shell` and tried `su`. Surprise, I have root on that device! At least on the shell, not in apps. So I moved that file in and restarted OM. No dice, it still wants me to download.
* Well… *Settings › Apps › OM:* force-close. Out there, back into the drawer, starting OM: :tada: Map accepted.

OK, need to remember those steps. Still, root required – I was lucky there as it's a test device with a test-keys ROM (hence root), others are mostly not. There should be a way without root – even if it's just OM importing ALL `*.mwm` files found and not just the two world maps.

So far, thanks for your help @biodranik – much appreciated and helping me a lot. If we had a non-root solution including "other maps", these steps should be written down. So far, for others searching it:

Use `adb shell` for the following steps:

1. go to `sdcard/Android/data/` and create a directory named `app.organicmaps` (if you want to use the F-Droid version)  or `app.organicmaps.web` (for the Github version).
2. inside that directory, create another one named `files`.
3. Place `World.mwm` and `WorldCoasts.mwm` into that `files` directory.
4. install and start OM, it should create a directory named with 6 digits and move the two files inside.
5. for additional maps, you need root access: copy them into that 6-digit directory.
6. force-close and restart OM, maps should be recognized now.

**Comment by biodranik** [comment]
I tried on the emulator and it worked without a root. World files are in /files/, map data files are in the corresponding folder with numbers.

Not sure about your permissions issue, maybe it depends on a device/ROM/Android version.

**Comment by IzzySoft** [comment]
Strange. Have you checked whom the directories are owned by in your emulator – and what permissions were set for owner, group, others? And was it Android-14 there as well?

**Comment by biodranik** [comment]
It was Android 15, an emulator from Android Studio. I didn't check permissions.

**Comment by IzzySoft** [comment]
Thanks! Well, I indeed can only speak for this one device (maybe others can confirm?). My other devices are still at Android < 12 and thus don't have that issue. And Emu often behaves slightly different from a real device…

**Comment by outusuke** [comment]
Add the option to save  maps directly to the Android/media folder instead of android/data . This would eliminate the need for workarounds like ADB or root access. Other apps like Kiwix and Antennapod already use this approach

**Comment by biodranik** [comment]
AFAIR, there is no fast way to directly access these maps in media folder from C++ code. Using slow map reads will slow down OM significantly.

**Comment by Wuzzy2** [comment]
I like to have this feature, too. I struggled for hours of making Organic Maps even work in the first place, as I wanted to install the maps without ever making Android touch the Internet.

But after following @IzzySoft’s advice <https://github.com/organicmaps/organicmaps/issues/4517#issuecomment-2287573601>, I found a way.

However, it took me dozens of attempts and re-tries until it finally worked. I don’t know what was the problem, maybe I made a mistake or something went wrong during data transfer. No idea. I also tried installing a later app version, so maybe that was it.

Anyway, may I suggest to add those step-by-step instructions to the official FAQ? Surely we two aren’t the only ones who want this.

For the FAQ, I suggest the following text (modified from @IzzySoft’s text):

> Q: How do I manually install maps without ever connecting Android to the Internet?
> 
> A: Having root access on your Android device is required, unless you only need the simplified World files with low detail.
> 
> First, you need to have a different device that has the `.mwm` files for the maps. To obtain those files, find a place where you can directly download map data for Organic Maps to your different device, like <https://omaps.webfreak.org/> or <https://omaps.wfr.software/>. Make sure that all map files you download are from the same date (in YYMMDD format).
> Second, download `World.mwm` and `WorldCoasts.mwm`. Then, download additional map files for the world regions you are interested in.
> 
> Once that is finished, follow these steps on your Android device:
> 
> 0. If Organic Maps is already installed, uninstall it
> 1. Go to `<root directory>/Android/data/` and create a directory named …
>     * … `app.organicmaps` (if you use the F-Droid version) or
>     * … `app.organicmaps.web` (for the GitHub version).
> 2. Inside that directory, create another one named `files`.
> 3. Copy `World.mwm` and `WorldCoasts.mwm` into that `files` directory.
> 4. Install and start Organic Maps (it should have created a directory named with 6 digits and moved the two files inside—check if this is the case).
> 5. For additional maps, you need root access: copy them into that 6-digit directory.
> 6. Force-close and restart Organic Maps (to force-close, go to Android’s Settings → Apps → Organic Maps → Force-close).
>
> If it worked, you now should have Organic Maps with functional maps. It should no longer ask you to download the world map on startup.

BTW, the "6-digit number" appears to be the YYMMDD date of the MWM files.
